File: python_src/heuristic.py
def evaluate_state(game, agent_color):

    """Evaluates the board from the agent's perspective using a simple heuristic"""
    
    #Check for win
    if game.winner == "white":
        return 100 if agent_color == "white" else -100
    
    elif game.winner == "black":
        return -100 if agent_color == "white" else 100
    
    elif game.is_draw:
        return 0  

    # Advancement heuristic
    white_max = 0  
    black_min = game.board.height 
    piece_advantage = 0  # metric for capturing pieces

    for (x, y), piece in game.board.board.items():
        if piece == "W":
            white_max = max(white_max, y)  
           
            if agent_color == "white":
                piece_advantage += 1  # if there are more white pieces, then it's better for white
        

        elif piece == "B":
            black_min = min(black_min, y) 
            
            if agent_color == "black":
                piece_advantage += 1  # if there are more black pieces, then it's better for black

    heuristic_value = ((black_min - 1) - (game.board.height - white_max)) + (piece_advantage * 10)

    # Adjust evaluation based on agent's perspective
    return heuristic_value if agent_color == "white" else -heuristic_value

File: cpp_src/AI/heuristic.cpp
#include "state.h"
#include "heuristic.h"
#include <algorithm>

#ifndef STATE_EVALUATION
#define STATE_EVALUATION

int evaluate_state(State& st, Player agent_color, Outcome outcome){
    if(outcome == WHITE_W)
        return agent_color == BLACK ? -100 : 100;

    if(outcome == BLACK_W)
        return agent_color == WHITE ? -100 : 100;

    if(outcome == DRAW)
        return 0;

    int white_max = 0;
    int black_min = st.get_height();
    int piece_advantage = 0;

    for(const auto& square : st.get_board()){
        if(square.second == WHITE){
            white_max = std::max(white_max, square.first.y);
            if(agent_color == WHITE)
                piece_advantage += 1;
        }
        else if(square.second == BLACK){
            black_min = std::min(black_min, square.first.y);
            if(agent_color == BLACK)
                piece_advantage += 1;
        }
    }

    float heuristic_value = ((black_min - 1) - (st.get_height() - white_max)) + (piece_advantage * 10);
    return heuristic_value * agent_color; 
}

#endif
